spyder/plugins/ipythonconsole/widgets/shell.py

```diff
@@ -454,18 +458,68 @@ def change_mpl_backend(self, command):
                 self.silent_execute(command)
 
     # ---- Spyder-kernels methods -------------------------------------------
+    def get_editor(self, filename=None):
+        """Get editor for filename and set it as the current editor."""
+        editorstack = self.get_editorstack()
+        if editorstack is None:
+            return None
+
+        if not filename:
+            return editorstack.get_current_editor()
+
+        index = editorstack.has_filename(filename)
+        if index is None:
+            return None
+
+        editor = editorstack.data[index].editor
+        editorstack.set_stack_index(index)
+        return editor
+
     def get_editorstack(self):
         """Get the current editorstack."""
         plugin = self.ipyclient.plugin
         if plugin.main.editor is not None:
             editor = plugin.main.editor
             return editor.get_current_editorstack()
+        raise RuntimeError('No editorstack found.')
 
     def handle_save_files(self):
         """Save the open files."""
+        self.get_editorstack().save()
+
+    def handle_run_cell(self, cell_name, filename=None):
+        """
+        Get cell code from cell name and file name.
+        """
         editorstack = self.get_editorstack()
-        if editorstack is not None:
-            editorstack.save()
+        editorstack.save()
+        editor = self.get_editor(filename)
+
+        if editor is None:
+            raise RuntimeError(
+                "File {} not open in the editor".format(filename))
+
+        editorstack.last_cell_call = (filename, cell_name)
+
+        # The file is open, load code from editor
+        return editor.get_cell_code(cell_name)
+
+    def handle_cell_count(self, filename=None):
+        """Get number of cells in file to loop."""
+        editorstack = self.get_editorstack()
+        editorstack.save()
+        editor = self.get_editor(filename)
+
+        if editor is None:
+            raise RuntimeError(
+                "File {} not open in the editor".format(filename))
```

Comment:
Again, do we need to raise an error here?

Reply:
If the call fails, the other side must be notified. Raising an error is the standard way of notifying the other side of an error.

Reply:
This error must be handled by the calling function. Here, if the calling function is `runcell`, it will simply print the error in the shell.

Reply:
Ok, no problem with this displaying this errors in the console.